blog/2024-01-25-inspecting-dependency-inference-results/index.mdx

```diff
@@ -0,0 +1,152 @@
+---
+    authors: [alexey]
+    tags: [concepts]
+---
+
+# Inspecting dependency inference results
+
+<CaptionedImg src={require("./splash.jpg").default}>
+  Photo by [Agence
+  Olloweb](https://unsplash.com/photos/magnifying-glass-near-gray-laptop-computer-d9ILr-dbEdg)
+  /
+  [Unsplash](https://unsplash.com/photos/magnifying-glass-near-gray-laptop-computer-d9ILr-dbEdg)
+</CaptionedImg>
+
+{/* truncate */}
+
+An important distinguishing feature of Pants v2 is its ability to automatically [infer your code's internal and external dependencies](https://www.pantsbuild.org/blog/2022/10/27/why-dependency-inference). The information about them is available via the `dependencies` and `peek` goals, so that metadata about what modules import from modules is always readily accessible.
+
+While managing a codebase in a language supported by Pants such as Python, Java, or Scala, it can, however, be useful to go further and explore your import statements to compare them with what has been inferred by Pants. This may be particularly helpful for large modules with lots of imports where it's easy to miss the fact that a dependency is omitted or that there's a dependency inferred that should really be ignored. To get help with that, you can use one of the less known goals, [`python-dump-source-analysis`](https://www.pantsbuild.org/2.18/reference/goals/python-dump-source-analysis), which can dump source code analysis for the Python source code build targets.
+
+This command may be used as a compliment to any static code analysis tooling you may already employ such as [semgrep](https://semgrep.dev/) or [flake8](https://flake8.pycqa.org/en/latest/) as they can only parse your modules semantically and often won't be able to link individual imports to build targets in your codebase.
+
+In order to run this goal, you need to enable the `"pants.backend.experimental.python"` backend:
```

Comment:
Why is it in experimental?

Reply:
Historical reasons, I guess - no one paid much attention to it...

Reply:
It hasn't been promoted to non-experimental. I think we'd want to standardise with the other goals at least for the default analysis flavour